### ISSUE TITLE
[yul] Should statements that affect the value returned by a subsequent call to `msize()` be removed even if they are redundant?

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#6827]
Main task here still to do:
 - document in description of optimizer and in changelog

----

## Description

(Found by libfuzzer+LPM)

```
{
let x,y
// After the computation of keccak256, the largest accessed memory index becomes 32 (0x20)
y := keccak256(1,1)
// msize() == 32
// Stores 32 at location 0
mstore(x, msize())
}
```

However, since `y` is unused it is removed by the optimizer. This means that the retained `mstore` now stores a zero (because `msize()` is zero).

Traces

```
Unoptimized trace
Trace:
Memory dump:
    20: 0000000000000000000000000000000000000000000000000000000000000020
Storage dump:

Optimized trace:
Memory dump:
Storage dump:
```

## Environment

- Compiler version: latest develop

**Comment by chriseth** [comment]
The old optimizer only removes `mload` and `keccak256` calls if the program does not contain the `msize` instruction. Maybe we should do the same here?

On the other hand, do people really rely on the result of `msize()` having a certain value instead of just the fact that `msize()` returns a pointer to prior unused memory? You can of course do something like

```
pop(mload(add(msize(), 20)))
```
and assume that this "allocates" 20 bytes, but I would argue that it exploits a quite peculiar feature of the EVM.

I mean we also do not guarantee that the value returned by `gas()` and `pc()` remains unchanged by the optimizer.

Is it enough to document this?

**Comment by chriseth** [comment]
Note that this will affect inline assembly, too!

**Comment by leonardoalt** [comment]
I would discourage that usage and optimize the statements away.

**Comment by axic** [comment]
The only real effect I can imagine is memory expansion cost would happen at a different part of the code and may cause revert at a different place (in case of out of gas), but that should have no effect on anything.

If we could somehow phase out the use of `gasleft` that would also reduce the number of contracts which depend on specific gas values.

**Comment by bshastry** [comment]
I think it is a good idea to document this and retain the current behavior (optimizing statements).

However, we need to think of a way to not trip the fuzzer if we decide to retain current behavior :-)

**Comment by axic** [comment]
The reason we added it to the optimise was that LLL relied on this behaviour. Perhaps we cannot change `libevmasm` because of that before we remove LLL.

**Comment by chriseth** [comment]
Ah right, it was because of LLL! Note that this is only about the yul optimizer.

**Comment by bshastry** [comment]
This issue does not reproduce on `0.5.8` so I would assume the "bug" is in unreleased Solidity.

**Comment by chriseth** [comment]
@bshastry yes, it is caused by the same thing that forced us to remove several log statements.

**Comment by chriseth** [comment]
Main task here still to do:
 - document in description of optimizer and in changelog

**Comment by chriseth** [comment]
We should maybe evaluate whether it is possible to have a behaviour like the old optimizer: Only optimize memory reads if there is no msize in the code. It is a little tricky with Solidity because yul code is usually optimized in isolation. We have multiple stages, though: During analysis, we can check for any msize instruction in the code and then not optimize memory if the user-supplied code has msize.

**Comment by chriseth** [comment]
An implementation of this would be:
 - create an AST walker that collects uses of `msize` (as builtin function call, and as functionalinstruction)
   This walker could later be re-used to check if a certain function has certain opcodes and which functions it calls -> #5608 
 - split the property `sideEffectFree` into `sideEffectFree` and `sideEffectFreeUpToMSize`
 - update the UnusedPruner to take a parameter that says if it can optimize ignoring msize.

**Comment by chriseth** [comment]
Ok, the above is a bit more complicated for solidity actually since we need to check all inline assembly blocks in all contracts and all libraries that have inline functions.

Maybe it would be better to just make that a yul optimizer setting and add an assertion that is triggered if the setting is active and the msize instruction is used.

**Comment by chriseth** [comment]
Solidity inline assembly:
 - provide optimizer settings to the type checker
 - report an error if yul optimization is activated and inline assembly uses `msize`.

Yul optimizer:
 - as outlined in the [above comment](https://github.com/ethereum/solidity/issues/6827#issuecomment-495570535)

**Comment by leonardoalt** [comment]
Closed by #6840

**Comment by Moony6789** [comment]
I have also problem with the same error with msize() instruction
![image](https://github.com/ethereum/solidity/assets/126175765/428f1a63-fe5c-4212-8ac0-ebb666583ee7)
![image](https://github.com/ethereum/solidity/assets/126175765/37fe2732-42c6-4d65-ab4a-54bd281040f5)
Could you solve this problem?
